include/circt/Dialect/SMT/SMTOps.td

```diff
@@ -16,8 +16,152 @@ include "mlir/IR/EnumAttr.td"
 include "mlir/IR/OpAsmInterface.td"
 include "mlir/Interfaces/InferTypeOpInterface.td"
 include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/Interfaces/ControlFlowInterfaces.td"
 
 class SMTOp<string mnemonic, list<Trait> traits = []> :
   Op<SMTDialect, mnemonic, traits>;
 
+def DeclareConstOp : SMTOp<"declare_const", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>
+]> {
+  let summary = "declare a symbolic value of a given sort";
+  let description = [{
+    This operation declares a symbolic value just as the `declare-const`
+    statement in SMT-LIB 2.6. The result type determines the SMT sort of the
+    symbolic value. The returned value can then be used to refer to the symbolic
+    value instead of using the identifier like in SMT-LIB.
+
+    The optionally provided string will be used as a prefix for the newly
+    generated identifier (useful for easier readability when exporting to
+    SMT-LIB). Each `declare_const` will always provide a unique new symbolic
+    value even if the identifier strings are the same.
+
+    Note that this operation cannot be marked as Pure since two operations (even
+    with the same identifier string) could then be CSEd, leading to incorrect
+    behavior.
+  }];
+
+  let arguments = (ins OptionalAttr<StrAttr>:$namePrefix);
+  let results = (outs Res<AnySMTType, "a symbolic value", [MemAlloc]>:$result);
+
+  let assemblyFormat = [{
+    ($namePrefix^)? attr-dict `:` qualified(type($result))
+  }];
+}
+
+def SolverOp : SMTOp<"solver", [
+  IsolatedFromAbove,
+  SingleBlockImplicitTerminator<"smt::YieldOp">,
+]> {
```

Comment:
Do you think it would make more sense to have a solver SSA Value instead, that is passed to the other operations (smt.check, or smt.assert for instance)? If I recall, this is the design you used to have in the draft PR?

I feel it makes the restriction "being called inside an `smt.solver`" more easier to check, as this is just an SSA value that is passed.

Or does this design makes it easier to lower to LLVM? Or what else does it simplify?

Reply:
The `smt.solver` does not only instantiate and limit the live-time of a solver that can be passed to the assert and check operations (like in the Z3 python API), but also the overall SMT context that stores all the config flags, nodes, reference counters, etc. Every operation in this dialect lowers to one (or multiple) calls to a Z3 C API function, all of which take this context as an argument. 

Having the region defining the SMT context's lifespan is quite important because when lowering to LLVM, you need to know when to create and initialize a context instance and when to deallocate it again. We could have a solver value provided as a special block argument in this region. But we still need the region checking that no solver value can enter or exit the region, otherwise it would be possible to mix contexts (pass SMT expression nodes built with different contexts to a constructor call for a new node) which would also be UB.

As you saw in the draft branch, we had this solver type (which we can lower to a struct containing a pointer to the context and another one to the solver) that was passed to the assert and check operations, which could be used to support multiple solvers next to each other. However, to solve the issue of guaranteeing that a context is available, we need this pointer to the context present at all operations. So, we could also lower all the SMT values (bool, bitvector, int, etc.) to a pair of pointers (one for the AST node, one for the context) which means we don't need to add the solver value to all operations, but only to the constant operations (bool, bitvector, int constants, declare_const, etc.) which don't get the pointer from an operand in the first place.

The current design basically targets lowering the context and solver pointers to LLVM globals. At the start of the solver region, we initialize them, and at the end, we deallocate them. All operations just access those globals, and if they follow the specification, they get a legal context; otherwise, they get an invalid pointer, thus UB.

Do you think we should add a solver/context type, add an implicit block argument to the solver region with this type, and require it as an operand for all operations that currently don't have an SMT value operand (i.e., constants)? This means we need to funnel the new solver value through all function calls and other control flow ops, i.e., slightly more complicated lowering, and have to add additional operands to the constant operations, but we can get rid of the globals, and get stronger guarantees.
Or leave it as is and have a somewhat easier lowering, require globals, and get fewer guarantees?

Reply:
For the record, @math-fehr @fabianschuiki and I discussed this offline and came to the conclusion that the current design and implementation make the most sense.